# Módulo 07, Tarefa 01

Vamos começar a mexer na nossa base de projeto? Já fizemos um exercício de montar a variável resposta, já pudemos perceber que essa atividade pode não ser trivial. Vamos agora trabalhar a base para que fique propícia ao *scikitlearn* para trabalharmos.

Lembrando, a base se chama demo01.csv, e originalmente está publicada [aqui](https://www.kaggle.com/rikdifos/credit-card-approval-prediction).

#### 1) Carregue a base e avalie:

- As variáveis
- Tipos de dados de cada variável
- Quantidade de missings
- Distribuição da variável resposta (mau)

In [31]:
import pandas as pd

import seaborn as sns
import matplotlib.pyplot as plt

In [98]:
df = pd.read_csv('/content/demo01.csv')
df.head(1)

sexo posse_de_veiculo posse_de_imovel  qtd_filhos tipo_renda  \
0    M                Y               Y           0    Working   

                        educacao estado_civil    tipo_residencia      idade  \
0  Secondary / secondary special      Married  House / apartment  58.832877   

   tempo_emprego  possui_celular  possui_fone_comercial  possui_fone  \
0       3.106849               1                      0            0   

   possui_email  qt_pessoas_residencia    mau  
0             0                    2.0  False

In [33]:
print(df.dtypes)

sexo                      object
posse_de_veiculo          object
posse_de_imovel           object
qtd_filhos                 int64
tipo_renda                object
educacao                  object
estado_civil              object
tipo_residencia           object
idade                    float64
tempo_emprego            float64
possui_celular             int64
possui_fone_comercial      int64
possui_fone                int64
possui_email               int64
qt_pessoas_residencia    float64
mau                         bool
dtype: object


In [34]:
df.isna().sum()

sexo                     0
posse_de_veiculo         0
posse_de_imovel          0
qtd_filhos               0
tipo_renda               0
educacao                 0
estado_civil             0
tipo_residencia          0
idade                    0
tempo_emprego            0
possui_celular           0
possui_fone_comercial    0
possui_fone              0
possui_email             0
qt_pessoas_residencia    0
mau                      0
dtype: int64

In [35]:
print(df['mau'].value_counts())

False    16260
True       390
Name: mau, dtype: int64


#### 2) Vamos montar um metadados

1. Crie um dataframe com os nomes de cada variável e o tipo de dados de cada variável.
2. Adicione uma coluna nesse *dataframe* chamada "qtd_categorias" e coloque nela o número de categorias correspondente de cada variável. 
    Dica: 
        1. inicie uma lista vazia
        2. faça um for ao longo dos nomes das variáveis, 
        3. conte o número de categorias dessa variável 
        4. acumule essa informação de 3. na lista que você criou em 1. 
        5. No final, essa lista pode ser atribuída à nossa variável.
3. Crie variáveis dummy para as variáveis necessárias (i.e. aquelas que são qualitativas e não estão armazenadas como {0, 1} ou {True, False}.

In [36]:
variaveis = ['sexo', 'posse_de_veiculo', 'posse_de_imovel', 'qtd_filhos',
             'tipo_renda', 'educacao', 'estado_civil', 'tipo_residencia',
             'idade', 'tempo_emprego', 'possui_celular', 'possui_fone_comercial',
             'possui_fone', 'possui_email', 'qt_pessoas_residencia', 'mau']

tipos_dados = [type(object).__name__, type(object).__name__, type(object).__name__,
               type(int).__name__, type(object).__name__, type(object).__name__,
               type(object).__name__, type(object).__name__, type(float).__name__,
               type(float).__name__, type(int).__name__, type(int).__name__,
               type(int).__name__, type(int).__name__, type(float).__name__, type(bool).__name__]

qtd_categorias = []

 #Logica para contar o numero de categorias da variavel
def contar_categorias(variavel):
    return len(df[variavel].unique())

In [37]:
for variavel in variaveis:
  qtd_categorias.append(contar_categorias(variavel))

In [38]:
dados = {'variavel': variaveis, 'tipo': tipos_dados, 'qtd_categorias': qtd_categorias}

df_metadados = pd.DataFrame(dados)

In [21]:
qtd_categorias

[2, 2, 2, 8, 5, 5, 5, 6, 5298, 3005, 1, 2, 2, 2, 9, 2]

In [ ]:
#Unindo os dataframes de tipos de dados e quantidade de categorias
df_metadados = pd.merge(tipos_dados, qtd_categorias, on='index')

In [99]:
variaveis_qualitativas = ['sexo','posse_de_veiculo','posse_de_imovel','tipo_renda', 'educacao', 'estado_civil', 'tipo_residencia']
df_dummies = pd.get_dummies(df[variaveis_qualitativas], drop_first=True)
df_dummies.head(1)

sexo_M  posse_de_veiculo_Y  posse_de_imovel_Y  tipo_renda_Pensioner  \
0       1                   1                  1                     0   

   tipo_renda_State servant  tipo_renda_Student  tipo_renda_Working  \
0                         0                   0                   1   

   educacao_Higher education  educacao_Incomplete higher  \
0                          0                           0   

   educacao_Lower secondary  educacao_Secondary / secondary special  \
0                         0                                       1   

   estado_civil_Married  estado_civil_Separated  \
0                     1                       0   

   estado_civil_Single / not married  estado_civil_Widow  \
0                                  0                   0   

   tipo_residencia_House / apartment  tipo_residencia_Municipal apartment  \
0                                  1                                    0   

   tipo_residencia_Office apartment  tipo_residencia_Rented apartment  \
0                                 0                                 0   

   tipo_residencia_With parents  
0                             0

In [103]:
df_metadados = pd.concat([df_metadados, df_dummies], axis=1)


#### 3) Crie variáveis dummy para as variáveis necessárias (i.e. aquelas que são qualitativas e não estão armazenadas como {0, 1} ou {True, False}. Crie um *dataframe* apenas com as variáveis apropriadas para entrada no scikitlearn - elimine as variáveis tipo *str*, mantendo apenas suas versões *dummy*.

In [105]:
dummy_data = pd.get_dummies(df, columns=['sexo','posse_de_veiculo','posse_de_imovel','tipo_renda', 'educacao', 'estado_civil', 'tipo_residencia'], drop_first=True)

In [112]:
numeric_data = dummy_data.select_dtypes(include=['number'])
numeric_data.columns

Index(['qtd_filhos', 'idade', 'tempo_emprego', 'possui_celular',
       'possui_fone_comercial', 'possui_fone', 'possui_email',
       'qt_pessoas_residencia', 'sexo_M', 'posse_de_veiculo_Y',
       'posse_de_imovel_Y', 'tipo_renda_Pensioner', 'tipo_renda_State servant',
       'tipo_renda_Student', 'tipo_renda_Working', 'educacao_Higher education',
       'educacao_Incomplete higher', 'educacao_Lower secondary',
       'educacao_Secondary / secondary special', 'estado_civil_Married',
       'estado_civil_Separated', 'estado_civil_Single / not married',
       'estado_civil_Widow', 'tipo_residencia_House / apartment',
       'tipo_residencia_Municipal apartment',
       'tipo_residencia_Office apartment', 'tipo_residencia_Rented apartment',
       'tipo_residencia_With parents'],
      dtype='object')

#### 4) Qual variável é mais poderosa?

Considere as variáveis ```possui_email``` e ```posse_de_veiculo```. Faça uma tabela cruzada entre elas e responda qual delas te parece mais poderosa para prever a probabilidade de ```mau = 1```?

In [119]:
tabela_cruzada = pd.crosstab(index=numeric_data['possui_email'], columns=numeric_data['posse_de_veiculo_Y'],)
print(tabela_cruzada)


posse_de_veiculo_Y     0     1
possui_email                  
0                   9341  5829
1                    837   643


Na tabela cruzada podemos perceber que a quantidade de pessoas que possuem veiculos mas nao possuem e-mails (5829) é muito maior que as pessoas que possuem e-mails mas nao possuem carros (837). 
Logo a variavel 'posse_de_veiculo' pode entregar mais informações.

#### 5) Salve a base, pois ela será utilizada no final deste módulo.

In [120]:
numeric_data.to_csv('metadados_demo01.csv', index=False)